### ISSUE TITLE
Image Loading and Image Dimensions

### ISSUE THREAD

**Issue by juanitogan** [flame-engine/flame#222]
It seems the docs and examples could use some clarification about how to use `loadAll` with images and audio.  I was attempting to avoid hard coding image dimensions when setting up the sprite rect but noticed the various sprite props (`originalSize`, `size`, `image`, `src`) that might have the dimensions were null directly after Sprite instantiation.

I had cached all images with `loadAll` as directed and as seen in various guides and projects.  But, after reasoning it out, it became clear that none are doing this correctly (or I'm missing something).  First, it became clear that there is no magical caching and this typically-seen methodology does not work:

```dart
import 'package:flame/flame.dart';

await Flame.images.loadAll(<String>["background.png", "title.png"]);
.
.
.
import 'package:flame/sprite.dart';

sprite = Sprite("title.png");
// Can't use sprite.originalSize.x etc...
```

Nope.  That second load of "title.png" is not cached.  After some digging through autocomplete, this appears to be the proper way, but I haven't seen it exampled anywhere:

```dart
import 'package:flame/flame.dart';

await Flame.images.loadAll(<String>["background.png", "title.png"]);
.
.
.
import 'package:flame/flame.dart';
import 'package:flame/sprite.dart';

sprite = Sprite.fromImage(Flame.images.loadedFiles["title.png"]);
// Can now use sprite.originalSize.x etc...
```

So, did I get it right?  Bummer to need the extra import in my sprite classes but, so be it.

---

Now, why I need my own rect instead of just setting sprite properties for size and position, I don't know -- but that's beside the point here.  Not sure what `Sprite.size` is for since it's not settable.  There is no `Sprite.position`.  Thus, setting up a separate `Rect myRect` and calling `mySprite.renderRect(c, myRect)` to handle a moving sprite instead of just setting `mySprite.position.x` and calling `mySprite.render(c)` seems like too much wiring to me.  (Or, maybe I should be using `SpriteComponent` instead of `Sprite` even though I wasn't looking for the extra features.)  But, that's just my opinion, and I think Flame is great so far.  Thank you for your work on it.

**Comment by luanpotter** [comment]
@juanitogan, thanks for your clear and well educated feedback!

I might need some more time to properly look into this, but I believe this my elaborate a little bit (my first thoughts).

Your first approach does use the cached image. However, it's fetched from the cache asynchronously, so you can't immediately access it. The second way you enforce it to be sync. We def. can improve, if not our apis, at leas the docs.

**Comment by juanitogan** [comment]
If it helps, here's my use case for auto sizing and scaling sprites on load:

```dart
class Title extends SpriteComponent {
    final MainLoop game;

    Title(this.game, double startY) {
        sprite = Sprite.fromImage(Flame.images.loadedFiles["title.png"]);
        x      = (game.screenSize.width - sprite.originalSize.x * game.scaleY) / 2;
        y      = startY * game.scaleY;
        width  = sprite.originalSize.x * game.scaleX;
        height = sprite.originalSize.y * game.scaleY;
    }
}
```

(I was hoping to use the HasGameRef mixin as well but, well, it isn't set until `add()` and thus is always null during construction.  Bugger.  Still learning the hard way.)

**Comment by juanitogan** [comment]
Just noting that I showed the use case because I'm only interested in solving the case of how to define sprites without hard coding all the dimensions.  This is primarily to allow the art to easily evolve but, also, I just don't like hard coding all those value regardless.

I can't say I care how I solve that case, I just wanted to solve it.  If there's a better way by redesigning a tangent feature, then, great, I'll take it.  Perhaps: `Sprite.fromCache()` ??

**Comment by erickzanardo** [comment]
> (I was hoping to use the HasGameRef mixin as well but, well, it isn't set until add() and thus is always null during construction. Bugger. Still learning the hard way.)

This also is something that bothers me, I was thinking on adding a new method for components, that is called by the BaseGame on the components after their initialization is done, what do you think, @luanpotter and @juanitogan ?

**Comment by juanitogan** [comment]
@luanpotter , What I ended up doing about my gameref needs in the constructor was to build features to get rid of the need.  In this particular case, I added a universal scaling feature to `Game` so that I no longer need to scale all child objects in a low-res game.  I tried limiting the feature to just `BaseGame` but then could only universally scale components.  That was not a bad design idea but ultimately I decided to go all the way to `Game` for broader coverage, and given the low impact.

I have some some lingering gameref needs outside the constructor, such as calls to Game-level methods like mute(), but I may move those as well since I seem to be shifting the design towards fewer Game-level properties and more globals.

----

Anyhow, just a "heads up" to all that I've been working on a large and breaking PR with this and other changes.  It's large because it mostly contains a refactor of Sprite and Animation construction, and such a refactor can't be done in small pieces.  I appreciate the work that got Flame this far but I see the growing pains and decided to just start rewriting a few things to make my game port from Qt to Flame go easier.  I suffered quite a bit of confusion ramping up with Flame (the port has now taken much longer than it took to write the original game in QtQuick, which was also my first Qt app) and I believe these changes will mitigate some of that.  I think I've rebuilt my early Flame game objects 4 times over by now trying to find the best practice.

While my Flame changes are breaking, they don't change the overall design pattern much.  I haven't gone that far yet even though I would like to see a different composite-game-entity design (the ComposedComponent mixin is good enough for me for now) and maybe an ECS like [this](https://github.com/sschmid/Entitas-CSharp).  What Flame calls a component, I call an entity or game object, and every base entity should be compositive of other entities.  The components would then be things like sprites, animations, and positions/transforms that make up the entity.  Sorry, I'm getting too far off topic now.

Anyway, I'm working on the doc changes now and hope to submit the PR shortly.  I have no idea if it will be liked or not, but it is what I'm going with for my game port.  I expect a few will hate on the changes but, if the PR is accepted, they can always re-add what I removed/changed if they really want it back.
